components/logins/src/error.rs

```diff
@@ -3,19 +3,49 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 use std::ffi::OsString;
+pub type Result<T> = std::result::Result<T, LoginsError>;
+// Functions which are part of the public API should use this Result.
+pub type ApiResult<T> = std::result::Result<T, LoginsStorageError>;
+
+pub use error_support::handle_error;
+use error_support::{ErrorHandling, GetErrorHandling};
+
+// Errors we return via the public interface.
+//
+// Named `LoginsStorageError` for backwards compatibility reasons, although
+// this name shouldn't need to be used anywhere other than this file and the .udl
+//
+// Note that there is no `Into` between public and internal errors, but
+// instead the `ErrorHandling` mechanisms are used to explicitly convert
+// when necessary.
+//
+// XXX - not clear that these actually need to use `thiserror`? Certainly
+// not necessary to use `#[from]` though.
+#[derive(Debug, thiserror::Error)]
+pub enum LoginsStorageError {
+    #[error("Invalid login: {0}")]
+    InvalidRecord(String),
+
+    #[error("No record with guid exists (when one was required): {0:?}")]
+    NoSuchRecord(String),
+
+    #[error("Encryption key is in the correct format, but is not the correct key.")]
+    IncorrectKey,
```

Comment:
Should keys with invalid formatting also be included with this one?  I think from the app's point of view both things require the same response.  If needed, there could be an inner enum to differentiate between the two cases.

Reply:
I did go around a few circles here and ultimately decided that InvalidFormat isn't really something the apps will need to handle tbh - it really just means "invalid json", and we expose methods to create the key. So the only time this could happen is if they decided to try and create the key themselves (which is just misuse) or if somehow the key was partially corrupted in their store (eg, trailing `}` lost or similar).

So this doesn't really seem like something apps would ever catch, and I don't think they do now. Or to put it another way, using error reporting to know this *does not* happen has value in its own right.